test/memory/memory_api_test.go

```diff
@@ -0,0 +1,191 @@
+//  Copyright (c) 2024 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package memory
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"go.fd.io/govpp/binapi/vpe"
+	"go.fd.io/govpp/test/vpptesting"
+	"runtime"
+	"runtime/metrics"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+)
+
+var apiNum = flag.Uint("api-num", 0, "Custom number API to test")
+
+// required go runtime metrics
+const (
+	heapBytesAllocs  = "/gc/heap/allocs:bytes"
+	heapBytesFrees   = "/gc/heap/frees:bytes"
+	heapObjectAllocs = "/gc/heap/allocs:objects"
+	heapObjectFrees  = "/gc/heap/frees:objects"
+	goroutines       = "/sched/goroutines:goroutines"
+)
+
+// TestAPIMemory tests the GoVPP memory consumption for various numbers of API calls
+func BenchmarkAPIMemory(b *testing.B) {
+	flag.Parse()
+
+	fmt.Printf("Running GoVPP API calls memory test\n\n")
+	test := vpptesting.SetupVPP(b)
+	vpeRPC := vpe.NewServiceClient(test.Conn)
+
+	samples := []metrics.Sample{
+		{Name: heapBytesAllocs},
+		{Name: heapBytesFrees},
+		{Name: heapObjectAllocs},
+		{Name: heapObjectFrees},
+		{Name: goroutines},
+	}
+	pass := true
+	testAPICalls := func(n uint, thresholds [3]uint64) {
+		tHolds0, tHolds1, tHolds2 := "N/A", "N/A", "N/A"
```

Comment:
If you want to make this even more generic, you could add thresholds for all the metrics, and by default set these thresholds to nil so that you could make the threshold checks and the formatting in a loop.

Reply:
Agreed, e.g. something like
```golang
ok := testAPICalls([]Metric{
  {Name: heapBytesAllocs, MaxValue: 2621440},
  {Name: heapBytesFrees, MaxValue: 3145728}, 
  {Name: heapObjectAllocs},
  {Name: heapObjectFrees},
  {Name: goroutines},
})
```